src/passes/RemoveNonJSOps.cpp

```diff
@@ -14,251 +14,137 @@
+    // operations to intrinsic calls while we're at it.
     if (!builder) builder = make_unique<Builder>(*module);
     PostWalker<RemoveNonJSOpsPass>::doWalkModule(module);
 
-    if (needNearestF32) {
-      module->addFunction(createNearest(f32));
-    }
-    if (needNearestF64) {
-      module->addFunction(createNearest(f64));
+    if (neededIntrinsics.size() == 0) {
+      return;
     }
-    if (needTruncF32) {
-      module->addFunction(createTrunc(f32));
-    }
-    if (needTruncF64) {
-      module->addFunction(createTrunc(f64));
-    }
-    if (needCtzInt32) {
-      module->addFunction(createCtz());
-    }
-    if (needPopcntInt32) {
-      module->addFunction(createPopcnt());
-    }
-    if (needRotLInt32) {
-      module->addFunction(createRot(RotLInt32));
-    }
-    if (needRotRInt32) {
-      module->addFunction(createRot(RotRInt32));
+
+    // Parse the wast blob we have at the end of this file.
+    //
+    // TODO: only do this once per invocation of wasm2asm
+    Module intrinsicsModule;
+    std::string input(IntrinsicsModule);
+    SExpressionParser parser(const_cast<char*>(input.c_str()));
+    Element& root = *parser.root;
+    SExpressionWasmBuilder builder(intrinsicsModule, *root[0]);
+
+    std::set<Name> neededFunctions;
+
+    // Iteratively link intrinsics from `intrinsicsModule` into our destination
+    // module, as needed.
+    //
+    // Note that intrinsics often use one another. For example the 64-bit
+    // division intrinsic ends up using the 32-bit ctz intrinsic, but does so
+    // via a native instruction. The loop here is used to continuously reprocess
+    // injected intrinsics to ensure that they never contain non-js ops when
+    // we're done.
+    while (neededIntrinsics.size() > 0) {
+      // TODO: move popcnt to the wast blob below and avoid special handling of
+      // it.
+      if (neededIntrinsics.erase(WASM_POPCNT32)) {
+        if (module->getFunctionOrNull(WASM_POPCNT32) == nullptr) {
+          module->addFunction(createPopcnt());
+        }
+      }
+
+      // Recursively probe all needed intrinsics for transitively used
+      // functions. This is building up a set of functions we'll link into our
+      // module.
+      for (auto &name : neededIntrinsics) {
+        addNeededFunctions(intrinsicsModule, name, neededFunctions);
+      }
+      neededIntrinsics.clear();
+
+      // Link in everything that wasn't already linked in. After we've done the
+      // copy we then walk the function to rewrite any non-js operations it has
+      // as well.
+      for (auto &name : neededFunctions) {
+        if (module->getFunctionOrNull(name) != nullptr) {
+          continue;
+        }
+        auto* curr = intrinsicsModule.getFunction(name);
+
+        std::stringstream newTypeNameS;
+        newTypeNameS << "importedType" << curr->type;
+        Name newTypeName(newTypeNameS.str().c_str());
+
+        auto* newType = module->getFunctionTypeOrNull(newTypeName);
```

Comment:
please add a helper method `ModuleUtils::copyFunction` which copies a function from one module to another.

if you're not sure what i mean, i could do that as a followup instead.